component/s3/file/s3.go

```diff
@@ -0,0 +1,223 @@
+	})
+
+	bucket, file := getPathBucketAndFile(args.Path)
+	s := &S3{
+		opts:       o,
+		args:       args,
+		health:     &component.Health{},
+		updateChan: make(chan []byte),
+		errorChan:  make(chan error),
+	}
+
+	w, err := newWatcher(bucket, file, s.updateChan, s.errorChan, args.PollFrequency, s3Client)
+	if err != nil {
+		return nil, err
+	}
+	s.watcher = w
+	return s, nil
+}
+
+// Run activates the content handler and watcher
+func (s *S3) Run(ctx context.Context) error {
+	go s.handleContentUpdate(ctx)
+	go s.watcher.run(ctx)
+	<-ctx.Done()
+	if s.cancel != nil {
+		s.cancel()
+	}
+
+	return nil
+}
+
+// Update is called whenever the arguments have changed
+func (s *S3) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+	if newArgs.PollFrequency <= time.Second*30 {
+		return fmt.Errorf("poll_frequency must be greater than 30s")
+	}
+
+	// TODO detect if args are different
+	s3cfg, err := generateS3Config(newArgs)
+	if err != nil {
+		return nil
+	}
+	s3Client := s3.NewFromConfig(*s3cfg, func(s3o *s3.Options) {
+		s3o.UsePathStyle = newArgs.Options.UsePathStyle
+	})
+
+	bucket, file := getPathBucketAndFile(newArgs.Path)
+
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	s.args = newArgs
+	s.watcher.updateValues(bucket, file, newArgs.PollFrequency, s3Client)
+
+	return nil
+}
+
+// CurrentHealth returns the health of the component
+func (s *S3) CurrentHealth() component.Health {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	return *s.health
+}
+
+func generateS3Config(args Arguments) (*aws.Config, error) {
+	configOptions := make([]func(*aws_config.LoadOptions) error, 0)
+	// Override the endpoint
+	if args.Options.Endpoint != "" {
+		endFunc := aws.EndpointResolverWithOptionsFunc(func(service, region string, _ ...interface{}) (aws.Endpoint, error) {
+			return aws.Endpoint{URL: args.Options.Endpoint}, nil
+		})
+		endResolver := aws_config.WithEndpointResolverWithOptions(endFunc)
+		configOptions = append(configOptions, endResolver)
+	}
+
+	// This incredibly nested option turns off ssl
+	if args.Options.DisableSSL {
+		httpOverride := aws_config.WithHTTPClient(
+			&http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: args.Options.DisableSSL,
+					},
+				},
+			},
+		)
+		configOptions = append(configOptions, httpOverride)
+	}
+
+	// check credentials
+	if args.Options.AccessKey != "" {
+		if args.Options.Secret == "" {
+			return nil, fmt.Errorf("if accesskey or secret are specified then the other must also be specified")
+		}
+		credFunc := aws.CredentialsProviderFunc(func(ctx context.Context) (aws.Credentials, error) {
+			return aws.Credentials{
+				AccessKeyID:     args.Options.AccessKey,
+				SecretAccessKey: string(args.Options.Secret),
+			}, nil
+		})
+		credProvider := aws_config.WithCredentialsProvider(credFunc)
+		configOptions = append(configOptions, credProvider)
+	}
+
+	cfg, err := aws_config.LoadDefaultConfig(context.TODO(), configOptions...)
```

Comment:
Will this support the environment variables the AWS SDK expects? Or do users have to configure the options block to authenticate?

Reply:
It supports profiles and env var, the AWS SDK handles that under the scenes.